torch/autograd/_functions/tensor.py

```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+
+        The second one is much more complex, but it doesn't
+        assume anything on the input. The main downside is
+        that it takes time O(n^2), where n = input.size(self.dim)
+        (i.e. the length of the cumulative product). This is in
+        contrast to the forward pass and the efficient algorithm,
+        which are both O(n).
+
+        The second algorithm is a simple application of the chain
+        rule. If x is an n-dimensional vector, and y = cumprod(x),
+        and F is the final cost, then
+
+        dF / dx_k = sum_j (dF / dy_j) * (dy_j / dx_k)   (1)
+
+        The term dF / dy_j is just grad_output[j] (assuming again
+        everything is one-dimensional).
+
+        The term (dy_j / dx_k) is easilly seen to be
+
+        if j >= k
+            dy_j / dx_k = prod_{1 <= i <= j, i != k} x_i
+        else:
+            dy_j / dx_k = 0
+
+        Note that the indicator (j>=k) can be taken out
+        by replacing the sum in (1) with a sum from
+        j = k to n.
+
+        Thus,
+        df / dx_k = sum_{k <= j <= n} grad_output[j] * (dy_j / dx_k)
+
+        with
+        dy_j / dx_k = prod_{1 <= i <= j, i != k} x_i     (2)
+
+        Note that this last term is just the cumulative product
+        with k omitted. Thus, if x_k (the input) is nonzero, we can
+        just express this as
+
+        dy_j / dx_k = (prod_{1 <= i <= j} x_i) / x_k
+                    = y_j / x_k
+
+        So therefore,
+
+        df / dx_k = sum_{k <= j <= n} grad_output[j] * y_j / x_k
+
+        so
+
+        grad_output = reverse_cumsum(grad_output * output) / input
+
+        If the input is nonzero, we need to calculate the dy_j / dx_k
+        by using the formula (2), called in the code omitted_products.
+
+        The way the code calculates it is simply by noting that
+
+        prod_{1 <= i <= j, i != k} x_i
+            = (prod_{1 <= i <= k} x_i) * (prod_{k + 1 <= i <= j} x_i)
+
+        the first term is calculated as prods_until_k, which since
+        doesn't depend in j is easy to vectorize.
+
+        The second term (indexed by j) is the cumulative product of
+        x_{k+1}, x_{k+2}, ..., x_n, and it's named in the code
+        prods_from_k_pkus_1, and it's calculated as a cumprod.
+
+        In order to vectorize this properly, we need to add to
+        omitted_products the dimensions where k > j, and therefore
+        dy_j / dx_k = 0, which is done right after the assert.
+        '''
+
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        if dim_size == 1:
+            return grad_output.clone()
+
+        #  Simple case with nonzero elements in the input
+        if (input != 0).all():
+            output = torch.cumprod(input, dim=self.dim)
+            return reverse_cumsum(output * grad_output, dim=self.dim) / input
+
+        if input.is_cuda:
+            idx = torch.cuda.LongTensor(range(dim_size))
+        else:
+            idx = torch.LongTensor(range(dim_size))
```

Comment:
Actually I think creating these tensors could be avoided (it's best to not initialize them with Python objects -- it's **slow**). The only places where you use them is `index_select`, but you only slice them, so a faster and equivalent way would be to do this:
```python
dim_padding = (slice(None, None),) * (self.dim - 1)
# now you can do this instead of input.index_select(self.dim, idx[k + 1:dim_size])
input[dim_padding + (slice(k+1, None),)]
```

This doesn't copy the memory and avoids the `index_select` kernel which is not a trivial operation.